### DISCUSSION TITLE
Why is Homebrew so slow?

### DISCUSSION THREAD

**Discussion by xiruizhao** [Homebrew/discussions#1177] [Everyday usage] [answered]
Non-network operations took one second!
- macOS 11.2.3 with Intel i7 2.2 GHz 
- /bin/bash 3.2.57(1)-release (x86_64-apple-darwin20)
- /bin/ruby  2.6.3p62 (2019-04-16 revision 67580) [universal.x86_64-darwin20]
```sh
hyperfine '/bin/bash /usr/local/bin/brew help' --warmup 1
# Benchmark #1: /bin/bash /usr/local/bin/brew help
#   Time (mean ± σ):     964.8 ms ±   7.0 ms    [User: 514.7 ms, System: 433.3 ms]
#   Range (min … max):   953.4 ms … 973.6 ms    10 runs
hyperfine '/bin/bash /usr/local/bin/brew list' --warmup 1
# Benchmark #1: /bin/bash /usr/local/bin/brew list
#   Time (mean ± σ):      1.151 s ±  0.031 s    [User: 622.7 ms, System: 509.2 ms]
#   Range (min … max):    1.125 s …  1.210 s    10 runs
```
Would using POSIX-compliant [dash](https://wiki.archlinux.org/index.php/Dash) (preinstalled on macOS) improve performance? `/usr/local/bin/brew` already declares POSIX compliance.

**Comment by SMillerDev** [comment]
It's busy, that is why it's slow. What kind of real world use-case do you have that needs speed optimization upon this?

**Reply by xiruizhao** [reply]
It's just annoying. Compare `pip`, which is 3-4x faster. (Non-network operations)
```sh
hyperfine 'python3.9 /usr/local/bin/pip help' --warmup 1
# Benchmark #1: python3.9 /usr/local/bin/pip help
#   Time (mean ± σ):     207.6 ms ±   1.4 ms    [User: 152.7 ms, System: 51.1 ms]
#  Range (min … max):   205.4 ms … 209.5 ms    14 runs


hyperfine 'python3.9 /usr/local/bin/pip list' --warmup 1
# Benchmark #1: python3.9 /usr/local/bin/pip list
#   Time (mean ± σ):     393.0 ms ±   2.6 ms    [User: 290.6 ms, System: 98.5 ms]
#   Range (min … max):   388.5 ms … 397.1 ms    10 runs
```

**Reply by SMillerDev** [reply]
Pip is a language package manager, it has different use cases and requirements then brew has as a system package manager.

**Reply by justdanyul** [reply]
> Pip is a language package manager, it has different use cases and requirements then brew has as a system package manager.

Try using apt-get , pacman, etc.

Long story short, brew is handy, but it sure is slow.

**Reply by gwerbin** [reply]
A good comparison would be MacPorts, which in extreme cases can download a full set of package updates *and* install one or two of them, in the same time it takes Brew to show `info` output for a Formula.

Also, the fact that Pip happens to be a language package manager doesn't make much of a difference in my mind. It performs broadly the same functions and its CLI has similar design constraints. And Pip is an interesting example because it's not particularly fast due to the overhead of starting Python.

**Reply by MikeMcQuaid** [reply]
@xiruizhao > `hyperfine 'python3.9 /usr/local/bin/pip list' --warmup 1`

Note when https://github.com/Homebrew/brew/pull/17317 is merged: `brew list` will be faster than `pip list`.

That's the real answer to "why is Homebrew slow": because more people are interested in complaining about Homebrew being slow than spending their evenings and weekends making it faster (like me).

**Reply by MikeMcQuaid** [reply]
And when https://github.com/Homebrew/brew/pull/17710 is merged `brew help` will be faster than `pip help`, too.

**Comment by alebcay** [comment, marked as answer]
> Would using POSIX-compliant dash (preinstalled on macOS) improve performance?

My guess would be probably not. `brew list` may be slower than it could be due to how some of the logic is overloaded - it's not used just for listing installed items, but also can filter by formulae only, casks only, unbrewed formulae, versions of a particular formula, pinned formulae only, etc.

Interestingly, from the numbers above, the CPU time for `brew` shows that `System` time makes up the majority of `User` time, while `pip` shows a `System` time that is about a third that of the `User` time.

You can find the code that drives `brew list` [here](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/list.rb). If you have a proposal on how to make the `brew list` command run faster, please kindly open a pull request with the requisite documentation and before/after benchmarks and the team would be more than happy to review and consider the change.

**Reply by gwerbin** [reply]
As I stated elsewhere, a much better comparison would be another package manager such as MacPorts, Pkgsrc/Pkgin, APT, Zypper, DNF, Pacman, XBPS, etc. Nearly all of these perform near-instant `list` filtering, `info` output, and other "offline" operations, and some of them support very complicated logic for filtering results.

Another interesting comparison is Conda, which is *also* famously slow, but its reimplementation Mamba is generally fast for offline operations such as listing packages, and the main bottleneck there is dependency resolution.

Looking at `brew list` specifically is maybe missing the forest for the trees a bit. There's something fundamental in the design of Brew that gives it a ton of CLI startup overhead, and generally makes commands slow. Does Brew use the filesystem differently? Does it lack certain layers of caching and recorded manifests that other package managers have? Etc.

**Reply by alebcay** [reply]
You raise good points regarding looking at the bigger picture. Nobody on the maintainer team that I know of is actively instrumenting/profiling `brew` holistically on a regular or rigorous basis.

For anyone that is interested in improving `brew` performance, the suggested course of action is still the same as in my original response. Of course, at a bigger picture scale, any sort of design change is subject to a number of other considerations aside from performance as well.

The `brew prof` command may be a helpful tool in collecting profiling/performance data. Run `brew prof --help` for more details.

**Comment by xiruizhao** [comment]
Ubuntu 20.04 with Intel J5005 CPU @ 1.50GHz
```sh
hyperfine 'apt list --installed'
# Benchmark #1: apt list --installed
#   Time (mean ± σ):     470.4 ms ±   4.4 ms    [User: 436.0 ms, System: 34.5 ms]
#   Range (min … max):   462.3 ms … 477.5 ms
```

**Comment by gwerbin** [comment]
I always assumed that Brew was slow because Ruby was slow. There are similar issues with CLI tools built in Python and NodeJS, where lots of little inefficiencies in the language runtime compound into really big overhead even for simple tasks. Compare NVM and Rbenv to FNM and Frum; it's night and day.

Apparently Ruby 3.0 is quite a bit faster, and there are also faster (albeit experimental) Ruby implementations that use either JIT compiling or traditional ahead-of-time compiling using Sorbet type annotations. Has anyone experimented with running Brew under those runtimes?

**Comment by rdp** [comment]
You could profile it to see why it's slow, then propose a fix. Here's one from me: convert code base to crystal (if it's cpu bound that is) ? :)

**Comment by Elly-Mandliel-Forter** [comment]
Homebrew is a piece of unusable software

**Comment by aaomidi** [comment]
It's the git fetch. If you're on a slow connection and haven't updated your git repo clone for a while, it's going to take a loooooong time.

**Reply by gwerbin** [reply]
The `brew` CLI command is slow to perform even basic non-network operations like `brew info`, which are nearly instant in other package managers.

**Reply by StephenLeafWW** [reply]
I noticed recently, while simply downloading a single package, that it opted to update my qemu, as well as my elixir stuff, as well as my erlang, as well as...... well it's updating everything.  I know not the reason why it decided that.  Maybe I did something wrong but I figured `brew install [package]` was an operation that wasn't going to be running for 45 min on a business line, 10gb/s down.

**Reply by gromgit** [reply]
That has been Homebrew policy for a long while now: https://docs.brew.sh/FAQ#why-does-brew-upgrade-formula-or-brew-install-formula-also-upgrade-a-bunch-of-other-stuff

**Comment by Loosie94** [comment]
Brew might be downloading something large.
Run for example: `brew update --debug --verbose`
To see its progress.

**Comment by fr0z3nfyr** [comment]
Well, Homebrew is indeed not among the fastest package managers, especially if any the operation network dependent (even if your corporate leased line is 10Gb/s. That's something the maintainers have to first admit to attempt a fix :disappointed:

Anyway, if someone stumbles across this thread when trying to perform network dependent operations (install, upgrade etc.) and wondering why brew is compiling almost everything, make sure the version of the OS is supported. [FAQ on this makes sense](https://docs.brew.sh/FAQ#why-do-you-compile-everything).

**Comment by cxpreng** [comment]
It's amazing how shitty a piece of software may get to be, after a few minutes, when I noticed brew was going nowhere "trying" to install php8.1, becaue stupid brew upgraded to 8.2 without asking me, and while it was downloading a lot of useless and non necessary "dependencies" I was able to download and install MacPorts and finish a php installation with it.
And I say brew was "trying" cause I finally got an error because of the macOS version, all good and working with MacPorts, brew I had it with you, you are going down, now I'm sure removing it is gonna be fun

**Reply by xiruizhao** [reply]
Compare [Ubuntu package libapache2-mod-php8.1 dependencies](https://packages.ubuntu.com/kinetic/libapache2-mod-php8.1) and [Homebrew php@8.1 formula dependencies](https://github.com/Homebrew/homebrew-core/blob/d2580a4baef2a8aa999f3a61b8b9421ab73f660c/Formula/php@8.1.rb), you can find many common dependencies. MacPorts probably bundles the dependencies in the package, but that is not the [standard packaging practice (as adopted by Debian and Fedora)](https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae).

**Reply by cxpreng** [reply]
Installing in Debian takes about 20% of the time brew takes, I do it all the time in servers with fewer resources than my MacBook has, and I correct, the error was because brew couldn't find Python which I also installed with brew 😂
Also apt won't try to install dependencies that are already installed as brew does

**Reply by gwerbin** [reply]
https://docs.brew.sh/FAQ#why-does-brew-upgrade-formula-or-brew-install-formula-also-upgrade-a-bunch-of-other-stuff

**Comment by abhi40308** [comment]
Just tried to install `watchman` and its been going on for 30 mins. Keeps updating dependency after dependency, almost feels like I just installed the next mac OS version. Funny to see people normalising this DX, giving all the reasons why this is necessary / unavoidable / configurable.

**Comment by cxpreng** [comment]
Months after removing this crap I still find orphan files scattered all around my system!! I did mention uninstalling was gonna be fun, and you guys did not disappoint!
You can't even write a decent uninstaller! 🤦🏻‍♂️
Oh the humanity!!

**Reply by gwerbin** [reply]
@cris-spreng my experience with Homebrew is that I can `rm -rf` the main directory (eg `/opt/homebrew`) and that's more or less a clean uninstall.

**Comment by dondxniel** [comment]
I literally just spent 5 hours trying to install and symlink node and npm (probably because my OS is outdated). And it still failed. After 5 hours.
There has to be a way to work around the speed constraint. Anyone knows how?

**Reply by SMillerDev** [reply]
Update your OS to a supported version so you can use prebuilt binaries again.

**Reply by cxpreng** [reply]
What is your OS?

**Comment by stuaxo** [comment]
On MacOS `brew --prefix` could do with being faster, I use it in two places in my `.zshrc` (one as `brew --prefix postgresql`) and get to watch the title run through, ruby and clang every new terminal.

**Reply by gromgit** [reply]
`brew --prefix <formula>` goes through all that to give you a value that you can count on, but if all you want is a reasonable starting point to drill down into a formula's files, then `${HOMEBREW_PREFIX}/opt/<formula>` is usually good enough and as fast as it gets. In the rare case when there's nothing there (renamed formula, perhaps), then run `brew --prefix <formula>` to get the canonical path...and consider caching it somewhere.

**Reply by stuaxo** [reply]
Cheers, I managed to get rid of those, I didn't know `$HOMEBREW_PREFIX` was a thing, and replaced my first use with it.

The second one I didn't need as I hadn't really considered all the binaries end up in `${HOMEBREW_PREFIX}/bin` as symlinks

Everything is faster now, with the rest of my slowdown not being homebrew at least.

**Reply by gromgit** [reply]
> The second one I didn't need as I hadn't really considered all the binaries end up in `${HOMEBREW_PREFIX}/bin` as symlinks

That assumption fails for keg-only formulae (e.g. `sqlite`). As a general rule, the public binaries for every formula can be found in `/opt/homebrew/opt/<formula>/bin`.

**Comment by rico-ocepek** [comment]
Sorry for the "me too" comment but...

...every time I have to use Homebrew I come back to this thread, while the installation is running, in the hopes that there's finally a fix for those truly slugish installation speeds 

I'm currently installing a small tool, that I wanted to try out quickly
`brew install scrcpy`

And 10 minutes later I'm still here looking at Homebrew downloading Mysql and ProtoBuf

<img width="764" alt="image" src="https://github.com/user-attachments/assets/f7e2884c-da6b-477e-9fc0-b0d9da122631" />

**Reply by SMillerDev** [reply]
Running https://github.com/DomT4/homebrew-autoupdate in upgrade mode would help with that, since the issue is that brew needs to catch up on the very outdated install. Brew works best if you regularly run `brew upgrade` to make sure it gets the latest versions of everything.

**Reply by rico-ocepek** [reply]
Thanks for the reply. This could work well for me in combination with what I stumpled upon today:

https://computingforgeeks.com/prevent-homebrew-auto-update-on-macos/

tl;dr it says you can disable updates during installation with `HOMEBREW_NO_AUTO_UPDATE=1 brew install <formula>` 

I will give this a try

Still I find the default behaviour of just updating every package I have when I just want to install something a bit weird

**Reply by SMillerDev** [reply]
> Still I find the default behaviour of just updating every package I have when I just want to install something a bit weird

https://docs.brew.sh/FAQ#why-does-brew-upgrade-formula-or-brew-install-formula-also-upgrade-a-bunch-of-other-stuff

**Reply by fr0z3nfyr** [reply]
and that breaks things... I have an old intel mac (running Monteray - ver. 12.7.5)... i tried installing a package and homebrew smoked up working packages to hell (I think I also did a self update for some reason) and now I'm stuck because some of the old packages are auto updated and some other packages vomited all errors on my face leaving partial installs (incompatible junk)... Brew very "gracefully" commented that the OS version is not supported anymore so it won't dance along... Utter nuisance

**Reply by p-linnane** [reply]
@fr0z3nfyr You're using an end-of-life OS, which is a [Tier 3 configuration](https://docs.brew.sh/Support-Tiers#tier-3). Upgrade your macOS if you want a better experience. There's no reason to come in here and get hostile because you've chosen to run unsupported software.

**Reply by Sarke** [reply]
> https://docs.brew.sh/FAQ#why-does-brew-upgrade-formula-or-brew-install-formula-also-upgrade-a-bunch-of-other-stuff

@SMillerDev This is why I left Windows years ago, because if you wanted to start it and do something it basically goes "fuck you, I'm busy" when it does it's updates and indexing and other nonsense in the background.

At the **BARE MINIMUM** it needs to have a overall progress bar to show how many packages it has installed out of how many, etc.  Currently it only shows a progress bar for downloading files, which this day in age is useless.  Things like this gets 10x more frustrating because we're X number of minutes in and we don't know if it's gonna be 1 or 5 or 15 more minutes. 

Perhaps even a warning to start, like apt, "this will install X, but also upgrade A, B, C, D, E, etc". Most users will wait for 30 min if they know it will take 30 minutes, but if something is just spinning it's wheels for God knows how long, it's that much more aggravating and we're just gonna take it out back and kill it.

**Reply by p-linnane** [reply]
> Most users will wait for 30 min if they know it will take 30 minutes, but if something is just spinning it's wheels for God knows how long, it's that much more aggravating and we're just gonna take it out back and kill it.

Upgrade your packages regularly and this isn't an issue. This only becomes a problem if you go for long stretches without upgrading.

**Reply by MikeMcQuaid** [reply]
> Perhaps even a warning to start, like apt, "this will install X, but also upgrade A, B, C, D, E, etc".

@Sarke Use `brew install --ask` if you want that messaging. Also, please re-read our Code of Conduct and adjust your tone in your future communication.

**Reply by Sarke** [reply]
> Upgrade your packages regularly and this isn't an issue. This only becomes a problem if you go for long stretches without upgrading.

While that may be true, it's not a good argument for why a basic progress indicator is a bad idea. I don't know a single other package manager that doesn't at least have a simple `38/47` type of progress indicator.

> Also, please re-read our Code of Conduct and adjust your tone in your future communication.

For the record, I was paraphrasing Windows from years ago.

**Reply by MikeMcQuaid** [reply]
> For the record, I was paraphrasing Windows from years ago.

@Sarke Noted. Regardless: please do so, thanks.

**Comment by akvachan** [comment]
Unfortunately brew is indeed extremely slow, it took me 3 hours to install 170 dependencies from scratch on a fresh OSX installation. The only solution forward is migrating my entire configuration to something like Nix, or building everything from scratch.

Edit: Homebrew is a terrible package manager, not only does it update itself every time you download something (by default), it is also painfully difficult to downgrade dependencies, it is extremely difficult to impossible to actually install anything within a hermetic environment. Unfortunately, there are not many alternatives...

**Reply by p-linnane** [reply]
3 hours to install 170 packages seems more like an issue with your network than `brew` itself.

Have you considered creating your own package manager? You seem to be very well versed in what package managers absolutely should do, so it may be worth looking in to.

**Reply by SEProblem** [reply]
While it's fun to be salty and passive aggressive, dismissing a perfectly fine critique; you shouldn't "gaslight" your users, it's dishonest and narcissistic.

The reality is that on my new work machine, not too long ago; downloading about 90 packages or so was an excessively long time, probably around 2 hours.  We have fiber here and the connection is solid, reliable and much faster than anything else I have access to outside of work.  The download speed was never the problem, it just seems that during heavy traffic the servers slow to a crawl and can't keep up.  If you want to slander the hardware, it's a m4 macbook pro, 48 GB of mem and pretty snappy I'd say.  Maybe I should have requested a nicer apple laptop since at an MSRP of around $3k it's a meager machine, definitely not fast enough to run brew.

I don't know what you gain from maintaining a project if you can't bear to hear issues with said project.  I get the sense that a lot of OSS developers are this way with any and all legit complaints.  Which is why I left OSS dev 10 years ago and never looked back and will never maintain any public facing project, the community is insufferable.

I still like brew, and it's fine enough for what it does, and I'm thankful for all of you keeping it going, but there are some brutal slowdowns.

This is a pain point shared by many users, you aren't more important than thousands of people just because you decided to maintain the project.

**Reply by p-linnane** [reply]
> While it's fun to be salty and passive aggressive, dismissing a perfectly fine critique; you shouldn't "gaslight" your users, it's dishonest and narcissistic.

It is not gaslighting to point out that multi-hour installs are almost always caused by local network conditions. Across enterprise fleets, fresh macOS installations, and many of my own systems, I have never seen the multi-hour behavior you are describing. Homebrew 5.0.0 also introduced parallel downloads, which makes things significantly faster.
If you can share logs or clear steps to reproduce the issue, we are happy to investigate it. A post that says “brew is terrible” does not give us anything we can act on.

> The download speed was never the problem, it just seems that during heavy traffic the servers slow to a crawl and can't keep up.

Our bottles are served through GitHub Packages. Microsoft’s infrastructure is not the limiting factor here. If you are consistently experiencing slow downloads, there is almost certainly something environmental involved. Reproducible data would help us confirm this.

> I don't know what you gain from maintaining a project if you can't bear to hear issues with said project. I get the sense that a lot of OSS developers are this way with any and all legit complaints. Which is why I left OSS dev 10 years ago and never looked back and will never maintain any public facing project, the community is insufferable.

There is a difference between actionable bug reports and vague complaints. When users provide details such as commands, logs, and diagnostics, we can help. When the starting point is “brew is terrible,” there is no specific information to address.

> I still like brew, and it's fine enough for what it does, and I'm thankful for all of you keeping it going, but there are some brutal slowdowns.

If you can provide steps to reproduce or diagnostics, we can look into it.

> This is a pain point shared by many users, you aren't more important than thousands of people just because you decided to maintain the project.

Nobody working on Homebrew is doing it for a sense of importance. We do it because the project helps a large number of people, and because our in-person interactions with users are overwhelmingly positive. If a small number of angry comments dictated how we worked, the project would not have survived for as long as it has.

**Comment by willkill07** [comment]
Why is `brew list` / `brew ls` so much faster than `brew info` or `brew search`?

I understand needing to parse JSON, but I have yet to see `brew info <pkg>` return in at less than 1.2s on an M3 Pro Macbook Pro. This seems painfully slow.

**Command:**
```console
time brew info gcc
```
<details>
<summary><b>Output:</b></summary>
<pre>
]==> gcc: stable 15.2.0 (bottled), HEAD
GNU compiler collection
https://gcc.gnu.org/
Installed
/opt/homebrew/Cellar/gcc/15.2.0 (1,602 files, 460.7MB) *
  Poured from bottle using the formulae.brew.sh API on 2026-01-01 at 16:39:34
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/g/gcc.rb
License: GPL-3.0-or-later WITH GCC-exception-3.1
==> Dependencies
Build: make ✘
Required: gmp ✔, isl ✔, libmpc ✔, mpfr ✔, zstd ✔
==> Options
--HEAD
        Install HEAD version
==> Analytics
install: 58,783 (30 days), 204,825 (90 days), 1,045,368 (365 days)
install-on-request: 31,675 (30 days), 112,617 (90 days), 562,444 (365 days)
build-error: 1,862 (30 days)
brew info gcc  0.55s user 0.20s system 60% cpu 1.238 total
</pre>
</details>

I also tried with a package with no dependencies to see if that had an effect:

**Command:**
```console
time brew info cmake
```

<details>
<summary><b>Output:</b></summary>
<pre>
==> cmake: stable 4.2.1 (bottled), HEAD
Cross-platform make
https://www.cmake.org/
Installed
/opt/homebrew/Cellar/cmake/4.2.1 (4,016 files, 63.2MB) *
  Poured from bottle using the formulae.brew.sh API on 2026-01-01 at 16:30:25
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/c/cmake.rb
License: BSD-3-Clause
==> Options
--HEAD
        Install HEAD version
==> Caveats
To install the CMake documentation, run:
  brew install cmake-docs
==> Analytics
install: 114,905 (30 days), 390,792 (90 days), 1,560,660 (365 days)
install-on-request: 83,276 (30 days), 289,072 (90 days), 1,194,538 (365 days)
build-error: 809 (30 days)
brew info cmake  0.62s user 0.19s system 65% cpu 1.245 total
</pre>
</details>

**Reply by MikeMcQuaid** [reply]
Run `brew prof info gcc` to find out. It's a profiler which will tell why it's slow so you can open a pull request and fix it.

> Why is `brew list` / `brew ls` so much faster than `brew info` or `brew search`?

Because it's done entirely in Bash because it doesn't need to read any Ruby. `brew info` does.

**Reply by willkill07** [reply]
> Run brew prof info gcc to find out.

Nifty! Why isn't this exposed from `brew --help`?

> so you can open a pull request and fix it.

As someone unfamiliar with the codebase, I feel like it's a bit of a stretch to have someone simply reporting an issue under this topic "Why is Homebrew so slow" **fix** it.

What if the underlying problem is parsing the massive API JSON every damn time? I can't "fix that" due to the breadth of scope and a major architectural change.

**Reply by MikeMcQuaid** [reply]
> Nifty! Why isn't this exposed from `brew --help`?

The same reason the other `brew commands` aren't: it's not a comprehensive list. It's in `man brew`.

> As someone unfamiliar with the codebase, I feel like it's a bit of a stretch to have someone simply reporting an issue under this topic "Why is Homebrew so slow" **fix** it.

Every single Homebrew maintainer at some point had no familiarity with the codebase, myself included.

You asked why it is slow: I provided the command you need to run to get that information.

> What if the underlying problem is parsing the massive API JSON every damn time? I can't "fix that" due to the breadth of scope and a major architectural change.

It's not. If you literally just ran the command I said to run you'd see it's ~13.31% of the runtime. 

Even if it was: you actually could indeed fix that same way as anyone else: one PR at a time.

Your profile says you "like making things go fast". I've spent a bunch of time and effort making Homebrew go faster. I'm sure it'd be faster if you spent some time an effort on it too.

Of course, you have every right to not do this but you seem to actually have some relevant career skills here so it would be a shame to simply complain about this and not make it better. The Homebrew maintainers are all volunteers here, it's no-one's job to make this faster.

**Reply by willkill07** [reply]
Work with a focus in high-performance C/C++/Rust/Python and distributed workflow optimization is very different than a Ruby-implemented client.

> Of course, you have every right to not do this but you seem to actually have some relevant career skills here so it would be a shame to simply complain about this and not make it better. The Homebrew maintainers are all volunteers here, it's no-one's job to make this faster.

And it's a shame that you are being so combative and hostile. Because this is why people don't want to work on OSS projects with maintainers who aren't very kind.

**Reply by MikeMcQuaid** [reply]
@willkill07 I used to do high-performance C++ work and the underlying optimisation principles are pretty similar.

I have invested a huge amount of time an energy into personally making this project better and mentoring pretty much every maintainer we've ever had in helping them to make it better too. Specifically around performance I've spent a bunch of time this week alone working on this.

With you specifically, I answered several questions you had and instead of saying "thanks" at any point you're going to call me "combative", "hostile" and "not kind". Ok, lesson learned: next time I'll ignore your questions because answering them seems to have been a mutual net negative.

> this is why people don't want to work on OSS projects with maintainers who aren't very kind.

Alternatively: this is why OSS maintainers don't want to answer questions that don't come from a place of "I want to help" but instead to criticise.

**Reply by willkill07** [reply]
Nifty is a way of expressing thanks.

This is a question forum, so I'm asking questions. Sorry to "misuse" this. I'll avoid being curious from now on 🫡

**Reply by willkill07** [reply]
You were the one who eagerly pushed me to make a contribution/PR in your very first response to me.

Me being curious and asking a question is very different than me wanting to work on an unfamiliar OSS project. I hope you can see that difference.

**Reply by MikeMcQuaid** [reply]
> Nifty is a way of expressing thanks.

Not one I've ever experienced. Future tip: if you're trying to say "thanks" and someone points out that they didn't think you did say "thanks", telling them (again) that they not doing things how you think they should also probably doesn't translate to "thanks".

Giving all my comments a 👎🏻 doesn't really communicate "thanks" either (looking forward to the 👎🏻 on this one, too).

> Me being curious and asking a question is very different than me wanting to work on an unfamiliar OSS project.

I do this to most people who seem to experience a problem with Homebrew that they have the capability to improve, both of which are the case for you. Many of them contribute, some of them are still maintainers. You can choose to be part of the solution or not.

I have more sympathy when people are entirely lacking in the necessary skills (e.g. they have never written a line of code before) and they have a blocking problem with Homebrew. Even some of those folks have written their first ever code in Homebrew.

**Comment by MikeMcQuaid** [comment]
I don't think any more conversation about "why is Homebrew slow" is going to help anyone at this point. I'm more than happy to help anyone with any PRs they want to open to help make Homebrew better. `brew prof` is the main tool you will find that will help you make Homebrew faster. Some of the specific cases in this discussion were addressed and we'll do our best to improve more in future.
